drivers/flash/flash_esp32.c

```diff
@@ -76,6 +87,227 @@ static inline void flash_esp32_sem_give(const struct device *dev)
 #include <stdint.h>
 #include <string.h>
 
+static int flash_esp32_read_check_enc(off_t address, void *buffer, size_t length)
+{
+	int ret = 0;
+
+	if (esp_flash_encryption_enabled()) {
+		LOG_DBG("Flash read ENCRYPTED - address 0x%lx size 0x%x", address, length);
+		ret = esp_flash_read_encrypted(NULL, address, buffer, length);
+	} else {
+		LOG_DBG("Flash read RAW - address 0x%lx size 0x%x", address, length);
+		ret = esp_flash_read(NULL, buffer, address, length);
+	}
+
+	if (ret != 0) {
+		LOG_ERR("Flash read error: %d", ret);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int flash_esp32_write_check_enc(off_t address, const void *buffer, size_t length)
+{
+	int ret = 0;
+
+	if (esp_flash_encryption_enabled() && !ENCRYPTION_IS_VIRTUAL) {
+		LOG_DBG("Flash write ENCRYPTED - address 0x%lx size 0x%x", address, length);
+		ret = esp_flash_write_encrypted(NULL, address, buffer, length);
+	} else {
+		LOG_DBG("Flash write RAW - address 0x%lx size 0x%x", address, length);
+		ret = esp_flash_write(NULL, buffer, address, length);
+	}
+
+	if (ret != 0) {
+		LOG_ERR("Flash write error: %d", ret);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+#if !defined(CONFIG_MCUBOOT) && DT_PROP(DT_CHOSEN(zephyr_flash), write_block_size) > 8
+#define FLASH_BUFFER_SIZE 32
+
+static bool aligned_flash_write(size_t dest_addr, const void *src, size_t size, bool erase);
+static bool aligned_flash_erase(size_t addr, size_t size);
+
+/* Auxiliar buffer to store the sector that will be partially written */
+static uint8_t write_aux_buf[FLASH_SECTOR_SIZE] = {0};
```

Comment:
Aren't those 2 buffers a waste of memory in case no flash encryption is used?

Reply:
I added some `ifdef` to keep a similar behavior as it was before when no Flash Encryption enabled, so the memory consumption may not increase for a regular use build.